### ISSUE TITLE
YDJ民科一句话读论文： GaussianSplatting & DreamGaussian[Splatting] & Dynamic/4D GaussianSplatting

### ISSUE THREAD

**Issue by yuedajiong** [graphdeco-inria/gaussian-splatting#335]
本人真的是民科(scientist-guise layman?)，正规的教育就中国西部一偏远农村上世纪初中毕业，所以瞎逼逼的，完全没有否定本文这种best paper先进性，反而是就当前很牛的算法在学习分析该方向的离想要的效果还有多远。 
（在外国朋友的项目下写中文，好像有点不太礼貌哈。）

A) 关于算法：
1）原始NeRF为了追求连续，搞得很慢，虽然光线用了采样。 hash和tensor-decom提速是另外一回事。
2）GaussSplat这里，抓点云，本质是在重点抓shape轮廓的关键点，这个就少了很多计算，而且确实还是关键点。 
   但问题是nerf采样还有机会抓住rgba上那些被pc丢掉的点。
   至少一个研究点是：如果colmap的pc相对于rgba纹理去比较，如果确实丢了很多关键点，得做一个colmap得修改产生pc++作为gausss输入点。文章这里大量频繁的tensor的操作（clone, remove等）
   感觉就图了一个快，自己的数据集上，效果提升没啥感觉。
   还有一个理解不知道对不对，点云做输入，都有了相当多的stereo的信息了。这不得算colmap-SfM（或者其他等价模块）的功劳吗？如果扣除这部分的效果，然后还能上best-paper?  （作者和论文都说了，论文中有相关的消融。 有sfm更好，不然有些部分容易模糊；我个人理解，有这个好初始值，不仅跟准，而且更快。）

B)关于代码：
这些大牛级的博士/科学家，做研究很厉害，虽然best paper，但代码写的真的不敢恭维。一堆的周边各种次要功能的代码(GUI, data processing, ...)，100+行的核心代码硬生生搞出几千行，好不容易才能撸出其中的关键模型和关键训练的部分。

用几行代码，给简化一下：
train():  #大写表示几个关键部件/步骤
    DATA: load cameras intrinsics & extrinsics  @./scene/dataset_readers.py  
    NETS: 算法点1-Network模型: create_from_pcd() @./scene/gaussian_model.py  #核心是pcd （其他ply都是恢复用，存为其他bin形式也行。其实那个ply用来存random/uniform的pc那个过程很快，完全没有必要写一个复杂的ply去又存又读的。开始训练的时候产生一个就好了；计算想要一样复现，一行代码控制一下seed就好了。）
  #由于这个算法的特点，所以不是常见的一个多层那种(比如参数去存priori)，它是从PointCloud的点/色做高斯化扩展表示后强拟合，所以参数也就集中在对应的点这里。个人理解，如果以后细节不够，需要增加点甚至粗细分阶段的优化，需要从这里改起。
        #主要这里：fused_point_cloud = torch.tensor(np.asarray(pcd.points)); fused_color = RGB2SH(torch.tensor(np.asarray(pcd.colors)); 和随后几行参数，opacity，R/t，scale
        #注：Scene @./scene/__init__.py 里面其实没啥，就存了点camera信息
    OPTIMIZER & SCHEDULER: 没有什么特殊的，Adam和简单控制了一下lr。
    LOOP:
        if iteration % 1000 == 0: gaussians.oneupSHdegree()  #increase SH-levels up to a maximum degree
        算法点2-GaussianRasterizer高斯栅格化渲染：render_pkg = render(viewpoint_cam, gaussians, pipe, background)  #render() @./gaussian_render/__init__.py
        算法点3-Densification致密化：gaussians.max_radii2D();  gaussians.add_densification_stats();  gaussians.densify_and_prune():clone+split+rune;  gaussians.reset_opacity()
        LOSS: alpah*L1 + (1-a)*SSIM

后面有空，搓个100行的代码丢上来。特别是不用SfM的PC做输入来优化。

2023-10-20:  路了一个300+行的版本，正确的。   额外还搞了一个mesh-extract的。 
参见：https://github.com/graphdeco-inria/gaussian-splatting/issues/350

**Comment by guanjunwu** [comment]
6

**Comment by ilovebing** [comment]
大佬，那您能告诉我他这个项目怎么导出来obj文件吗，带有纹理和颜色的。现在生成的Ply文件就是一堆相同颜色的点云，根本用不了。用blender或者meshlab打开转换格式报错没有faces

**Comment by liuzhao6322** [comment]
> 大佬，那您能告诉我他这个项目怎么导出来obj文件吗，带有纹理和颜色的。现在生成的Ply文件就是一堆相同颜色的点云，根本用不了。用blender或者meshlab打开转换格式报错没有faces
 可以到导入到ue5 里面  但是需要在虚幻商城里下载 3D Gaussian plug插件才能导入ply。导出obj显然是不可能的。导出去也不能用，需要自己写

**Comment by yuedajiong** [comment]
楼上两个兄弟 @ilovebing  @liuzhao6322   请参看这里：  https://github.com/graphdeco-inria/gaussian-splatting/issues/350

另外，我对高斯溅射，是这么理解的。  从作用上，他其实和NeRF一样，通过多图获得了以一种半隐式的表示，NeRF藏在MLP种，GS藏在这个扩展点云中。  这个原始论文，并没有提取mesh的工作，现在我看到的是DreamGauss这论文里面有，就是有个提取函数，用mcubes来构造.

当然，GS这种算法，其实提供了另外一种可能，就是我们未必要真的表示为显式的mesh了。  本来头发，水波等很多东西也不适合表示为mesh。 就算普通物体，弄成mesh最终也是为了栅格化渲染。  它这个每秒60+帧，以后只会更快。

现在DreamGauss，这篇论文，除了extract mesh这个，其实他主要工作：一是引入zero123来支持单图先验，二是在初始的mesh后面做了第二阶段优化。

长远来看，我觉得单图这个部分，甚至结合pose预测，更有价值。

**Comment by yuedajiong** [comment]
我把他代码修改重写为300+行了，还有一个整理的mesh功能。

**Comment by kangpeilun** [comment]
> 我把他代码修改重写为300+行了，还有一个整理的mesh功能。

大佬，你能开源你重写后的版本么？我想瞅瞅您怎么搞的，还有可微渲染那部分您也重写了？

**Comment by yuedajiong** [comment]
兄弟，渲染在重写。  他垃圾代码太多了。 

1， 我准备把backgroud的合并提取出来，整个论文的代码时强制在c/cu中合成，那么必须要一个背景分割的独立步骤。
2.   我在看，要不要实现batch操作。
3.   最终这东西，简单的render不够，viewer也不够，我最终想实现为gauss-point based editor。

上午在仔细研读论文： https://github.com/graphdeco-inria/gaussian-splatting/issues/392

我修改好了，会分享。

**Comment by kangpeilun** [comment]
阿里嘎多




康佩伦
***@***.***




------------------ 原始邮件 ------------------
发件人: yuedajiong ***@***.***&gt;
发送时间: 2023年10月27日 12:00
收件人: graphdeco-inria/gaussian-splatting ***@***.***&gt;
抄送: KangPeiLun ***@***.***&gt;, Comment ***@***.***&gt;
主题: Re: [graphdeco-inria/gaussian-splatting] YDJ民科一句话读论文： GaussianSplatting &amp; DreamGaussian[Splatting] &amp; Dynamic/4D GaussianSplatting (Issue #335)





 
兄弟，渲染在重写。  他垃圾代码太多了。
 
1， 我准备把backgroud的合并提取出来，整个论文的代码时强制在c/cu中合成，那么必须要一个背景分割的独立步骤。
 2.   我在看，要不要实现batch操作。
 3.   最终这东西，简单的render不够，viewer也不够，我最终想实现为gauss-point based editor。
 
上午在仔细研读论文： #392
 
我修改好了，会分享。

**Comment by gootyking** [comment]
> 楼上两个兄弟 @ilovebing @liuzhao6322 请参看这里： #350
> 
> 另外，我对高斯溅射，是这么理解的。 从作用上，他其实和NeRF一样，通过多图获得了以一种半隐式的表示，NeRF藏在MLP种，GS藏在这个扩展点云中。 这个原始论文，并没有提取mesh的工作，现在我看到的是DreamGauss这论文里面有，就是有个提取函数，用mcubes来构造.
> 
> **当然，GS这种算法，其实提供了另外一种可能，就是我们未必要真的表示为显式的mesh了。 本来头发，水波等很多东西也不适合表示为mesh。 就算普通物体，弄成mesh最终也是为了栅格化渲染。 它这个每秒60+帧，以后只会更快。**
> 
> 现在DreamGauss，这篇论文，除了extract mesh这个，其实他主要工作：一是引入zero123来支持单图先验，二是在初始的mesh后面做了第二阶段优化。
> 
> 长远来看，我觉得单图这个部分，甚至结合pose预测，更有价值。


先给大神跪一个~
我可能理解的不深哈，有个想问的问题请指教：如果不考虑导出mesh，直接就渲染每秒60帧，这个还是基于PC上强大的GPU吧，如果要跑在手机端，甚至手机的web端，能不能跑的动，可以的话有啥方向可以做出渲染的引擎不.... 谢谢大神

**Comment by kwea123** [comment]
https://github.com/MrNeRF/awesome-3D-gaussian-splatting#open-source-implementations
這裡有一大堆別人寫好的，各種平台都能很快

**Comment by gootyking** [comment]
> https://github.com/MrNeRF/awesome-3D-gaussian-splatting#open-source-implementations 這裡有一大堆別人寫好的，各種平台都能很快

谢谢大佬！我去学习学习
